Movie (Game Inputs) recording and playback

danzel opened the pull request:
Heavily inspired by the [Dolphin implementation](https://github.com/dolphin-emu/dolphin/blob/master/Source/Core/Core/Movie.h).

Adds command lines to the SDL2 exe to control playback and recording. We should add something to the QT UI to match in the future.
(Is there somewhere these should be documented other than just in the `--help` response?)

Every time the hardware reads an input device (buttons, touch screen, accelerometer, gyroscope, c-stick) we copy (when recording) or replace (when playing) the values.
Once playback has finished, we release control of inputs so that the user can take control.

It is possible (likely even) that replays will desync between versions of Citra as behaviours change. This is expected.

File format will need documenting in the wiki. I've hopefully left enough padding room to populate the header with other fields we want in the future, I've brought over a minimal set to start with.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/citra-emu/citra/2882)
<!-- Reviewable:end -->

lioncash commented on src/core/movie.cpp:
```diff
@@ -0,0 +1,421 @@
+        FileUtil::IOFile save_record(Settings::values.movie_record, "wb");
+
+        CTMHeader header;
+        memset(&header, 0, sizeof(CTMHeader));
```
You can just do `CTMHeader header = {};` to zero it out.

lioncash commented on src/core/movie.cpp:
```diff
@@ -0,0 +1,421 @@
+        std::string rev_bytes;
+        CryptoPP::StringSource(Common::g_scm_rev, true,
+                               new CryptoPP::HexDecoder(new CryptoPP::StringSink(rev_bytes)));
+        memcpy(header.revision, rev_bytes.data(), 20);
```
```cpp
memcpy(header.revision, rev_bytes.data(), sizeof(CTMHeader::revision));
```

wwylele replied:
and `std::memcpy`

lioncash commented on src/core/movie.cpp:
```diff
@@ -0,0 +1,421 @@
+}
+
+void Shutdown() {
+    if (play_mode == PlayMode::Recording) {
```
You can do:

```cpp
if (!IsRecordingInput()) {
    return;
}
```

to unindent this block of code.

lioncash commented on src/core/movie.cpp:
```diff
@@ -0,0 +1,421 @@
+        struct {
+            u16 x;
+            u16 y;
+            bool valid;
```
Note that `bool` isn't required to have a fixed bit size in the standard, so you may want to use a different type if you're writing it out to disk.

lioncash commented on src/core/movie.cpp:
```diff
@@ -0,0 +1,421 @@
+    if (!Settings::values.movie_play.empty()) {
+        LOG_INFO(Movie, "Loading Movie for playback");
+        FileUtil::IOFile save_record(Settings::values.movie_play, "rb");
+        u64 size = {save_record.GetSize()};
```
Why the braces?

lioncash commented on src/citra/citra.cpp:
```diff
@@ -55,6 +57,9 @@ int main(int argc, char** argv) {
     int option_index = 0;
     bool use_gdbstub = Settings::values.use_gdbstub;
     u32 gdb_port = static_cast<u32>(Settings::values.gdbstub_port);
+    std::string movie_record = "";
```
`= "";` isn't necessary. `std::string` constructs as empty by default.

lioncash commented on src/citra/citra.cpp:
```diff
@@ -121,11 +131,17 @@ int main(int argc, char** argv) {
     // Apply the command line arguments
     Settings::values.gdbstub_port = gdb_port;
     Settings::values.use_gdbstub = use_gdbstub;
+    Settings::values.movie_play = movie_play;
```
These strings can be `std::move`d.

lioncash commented on src/core/hle/service/ir/ir_rst.h:
```diff
@@ -9,6 +9,18 @@
+union PadState {
+    u32_le hex{};
+
+    BitField<14, 1, u32_le> zl;
```
`common/bit_field.h` needs to be included. As does `common/common_types.h`

lioncash commented on src/core/movie.cpp:
```diff
@@ -0,0 +1,421 @@
+#include <vector>
+#include <cryptopp/hex.h>
+#include "common/bit_field.h"
+#include "common/file_util.h"
```
`<cstring>`, `<string>`, and `common/common_types.h` needs to be included.

lioncash commented on src/core/hle/service/ir/ir_rst.cpp:
```diff
@@ -12,23 +12,12 @@
-union PadState {
-    u32_le hex{};
-
-    BitField<14, 1, u32_le> zl;
```
The `common/bit_field.h` header include can be removed.

danzel commented:
Thanks @lioncash, done everything except moving Movie in to a class.
I can do, just need a bit of advice on how it should look.
I think the public API (movie.h) stays as it does now, then inside we move all of the static variables and methods in to a new class that has a singleton instance created when `Init()` is called.
Sound right?

wwylele commented on src/core/movie.cpp:
```diff
@@ -0,0 +1,474 @@
+
+static void Play(Service::HID::PadState& pad_state, s16& circle_pad_x, s16& circle_pad_y) {
+    ControllerState s;
+    memcpy(&s, &temp_input[current_byte], sizeof(ControllerState));
```
`std::memcpy`. Same for below

wwylele commented on src/core/movie.cpp:
```diff
@@ -0,0 +1,474 @@
+    header.filetype[0] = 'C';
+    header.filetype[1] = 'T';
+    header.filetype[2] = 'M';
+    header.filetype[3] = 0x1B;
```
"CTM\x1B` should be defined as a constant, shared with the one used in `ValidateHeader`

wwylele commented on src/core/movie.h:
```diff
@@ -0,0 +1,42 @@
+void HandleCirclePad(Service::IR::CirclePadResponse& circle_pad);
+
+// TODO: Handle C-Stick
+}
```
new line at the end

wwylele commented on src/core/movie.h:
```diff
@@ -0,0 +1,42 @@
+
+void Shutdown();
+
+void HandlePadAndCircleStatus(Service::HID::PadState& pad_state, s16& circle_pad_x,
```
Need some documentation of what these `HandleSomething` do

danzel commented:
Thank you wwylele. Have implemented all of your feedback hopefully.

I renamed the fields to `ir_rst` and `extra_hid_response` to line them up with what the structs they are copying are called.

When making the `GetStickDirectionState` related change I discovered I was not storing the `circle_x/y` and `ir_rst.x/y` values correctly. So that was fixed also.


In dolphin they have a flag that controls if the system is trying to run in a deterministic mode.
ref: https://github.com/dolphin-emu/dolphin/blob/master/Source/Core/Core/Core.cpp#L959

We could look to implement something similar to control parts of the engine that may be non-deterministic for performance.

lioncash commented:
> I think the public API (movie.h) stays as it does now

@danzel Note that by keeping it wholly namespace-based externally makes it more difficult to get to the point where more than one instance of the emulator can run in the same process (which is something that should be allowable from a library point of view, IMO). Ideally you'd be able to just create a new System object and be fine (the only reason it's a singleton right now is because other code and globals haven't been amended yet).

lioncash commented on src/core/movie.cpp:
```diff
@@ -0,0 +1,461 @@
+        LOG_INFO(Movie, "Loading Movie for playback");
+        FileUtil::IOFile save_record(Settings::values.movie_play, "rb");
+        u64 size = save_record.GetSize();
+        save_record.Seek(0, SEEK_SET);
```
`GetSize()` seeks back to the original position it was at before the call, in this case, so this Seek() is likely not required.

lioncash commented on src/core/movie.cpp:
```diff
@@ -0,0 +1,461 @@
+
+    CTMHeader header = {};
+
+    std::memcpy(header.filetype, header_magic_bytes, 4);
```
You can make the filetype member initialize to the magic bits automatically as part of the initialization if you utilize `std::array`, like so:

```cpp
constexpr std::array<u8, 4> header_magic_bytes{{'C', 'T', 'M', 0x1B}};

#pragma pack(push, 1)
struct CTMHeader {
    std::array<u8, 4> filetype = header_magic_bytes;  // Unique Identifier (always "CTM"0x1B)
    u64 program_id;  // Also called title_id
    std::array<u8, 20> revision; // Git hash

    std::array<u8, 224> reserved; // Make heading 256 bytes, just because we can
};
static_assert(sizeof(CTMHeader) == 256, "CTMHeader should be 256 bytes");
#pragma pack(pop)
```

or alternatively the same as above can be done, minus the assignment in the struct and instead just do the assignment at the site it's needed at. like:

```cpp
...

CTMHeader header = {};
header.filetype = header_magic_bytes;

...
```

lioncash commented on src/core/movie.cpp:
```diff
@@ -0,0 +1,461 @@
+        return;
+    }
+
+    LOG_INFO(Movie, "Saving movie");
```
You may want to move the saving code to a `SaveMovie()` function to keep the logic separate from the shutdown logic if it ever needs to change. You'll likely also want to check if the file status is still good (in case the attempt to open the file fails) before continuing onwards attempting to save data.

lioncash commented on src/core/movie.cpp:
```diff
@@ -0,0 +1,461 @@
+}
+
+static bool ValidateHeader(const CTMHeader& header) {
+    if (std::memcmp(header_magic_bytes, header.filetype, 4) != 0) {
```
this can be simplified to `if (header_magic_bits != header.filetype)` if `std::array` is used as mentioned in my other comments.

lioncash commented on src/core/movie.cpp:
```diff
@@ -0,0 +1,461 @@
+    s.extra_hid_response.zr_not_held.Assign(extra_hid_response.buttons.zr_not_held);
+
+    temp_input.resize(current_byte + sizeof(ControllerState));
+    memcpy(&temp_input[current_byte], &s, sizeof(ControllerState));
```
This one's missing the `std::` prefix.

lioncash commented on src/core/movie.cpp:
```diff
@@ -0,0 +1,461 @@
+    ControllerStateType type;
+
+    union {
+        struct {
```
Should some of the `u16`/`s16` members have `_le` prefixes as well?

wwylele replied:
> Checked over the s16/u16, they don't need _le, only extra_hid_response has it in its source.

This is not how it works. All members specified for a file format should be tagged with endianess _for platform compatibility_ (as you wrote in the comment), which is independent from the endianess of the data source.

lioncash commented on src/core/movie.cpp:
```diff
@@ -0,0 +1,461 @@
+#pragma pack(push, 1)
+struct CTMHeader {
+    u8 filetype[4];  // Unique Identifier (always "CTM"0x1B)
+    u64 program_id;  // Also called title_id
```
Should this be a `u64_le`?

danzel replied:
ReadProgramId is: `virtual ResultStatus ReadProgramId(u64& out_program_id)`

lioncash commented on src/core/movie.cpp:
```diff
@@ -0,0 +1,461 @@
+#include "common/common_types.h"
+#include "common/file_util.h"
+#include "common/logging/log.h"
+#include "common/scm_rev.h"
```
`common/swap.h` needs to be included for the `_le`/`_be` types.

lioncash commented on src/core/hle/service/ir/ir_rst.h:
```diff
@@ -4,11 +4,25 @@
 #pragma once
 
+#include "common/bit_field.h"
+#include "common/common_types.h"
```
`common/swap.h` needs to be included for the `_le`/`_be` types.

lioncash commented on src/core/hle/service/ir/extra_hid.h:
```diff
@@ -6,12 +6,29 @@
 
 #include <array>
 #include <atomic>
+#include "common/bit_field.h"
```
`common/swap.h` needs to be included for the `_le`/`_be` types.

lioncash commented on src/core/movie.cpp:
```diff
@@ -0,0 +1,461 @@
+    }
+}
+
+void Shutdown() {
```
shouldn't this function, as the API currently is, reset `current_byte` to `0`?

danzel commented:
Checked over the s16/u16, they don't need `_le`, only `extra_hid_response` has it in its source.

Thanks for the `std::array` advice, that's much nicer, used them all :)

I agree this should be an object living inside system for the reasons you've mentioned.
Right now most of the parts of the system are violating this though I think.
https://github.com/danzel/citra/blob/912b3cbdbe99053956228751437f553ca0e381a6/src/core/core.cpp#L151-L157
Happy to work on this in the future as a separate item.

danzel commented:
Moved the duplicated code from the end of the Record methods into another method to get rid of the duplication and Squashed the commits.

Can see the (mostly original) history here:
https://github.com/danzel/citra/commits/movie-squash-no

Tried making it flush the inputs to disk every time we record one, but (as expected) it completely kills framerate. This would be nice for recording lockups, may need revisiting in the future.

B3n30 commented:
What's the status of this PR?

danzel commented:
Rebased to fix merge issue, should be good to go again once the builds run.

danzel commented:
Oh I see! ok will fix that shortly, thanks.

danzel commented:
Ok, endianness specified :+1:

Subv commented:
@wwylele Async file IO doesn't have to be completely non-deterministic, we can provide a fixed (and deterministic for each file operation) time window to perform the IO operation and block the emulation thread if it takes longer than that.

I'll give this a review either today or tomorrow.

Subv commented:
@danzel Does the recording and playback happen per-frame? Could it be possible to implement a playback system in a real 3DS using the same movie files this PR accepts?

danzel commented:
Recording and playback happens at the hardware read rate, which is more than once per frame.
https://github.com/danzel/citra/blob/a6cd3a9e5ddc537c519120ebe8418659a057ff3f/src/core/hle/service/hid/hid.cpp#L50-L53

We could change it to be once per frame, but that would require changing all of the controller input to be updated once per frame. SDL does this currently for keyboard/mouse inputs (not joystick). QT does not do this at all.
I considered changing all user controller inputs to be updated once per frame, but wylele didn't like it, so I've left it as it is. I don't think it is important to change.

As for playing these back on Hardware... I've done close to zero 3ds stuff, so I'm not sure.
Assuming we can hook the relevant sections that read the hardware and redirect them to get the values from this file, then yeah it should be possible.
I can look at this in the future if someone can point me in the right direction.

Fixing the clang-format now...

Subv commented on src/core/movie.cpp:
```diff
@@ -0,0 +1,470 @@
+static void Play(Service::HID::TouchDataEntry& touch_data) {
+    ControllerState s;
+    std::memcpy(&s, &temp_input[current_byte], sizeof(ControllerState));
+    current_byte += sizeof(ControllerState);
```
Why does this copy and skip the entirety of ControllerState if only TouchData was requested?

Is the current system dependent on the update order of the different devices?

danzel replied:
ControllerState is a union of all of the different inputs recordings we take, so it will be a `touch` or a `pad_and_circle` etc, it doesn't have both (if I got my `union`ing correct).

The system is dependent on the update order of the different devices. It will break if this ever changes.
However this should only change if we find the update rate is different than we have specified (the update rate is based on hardware tests)

Subv commented on src/core/movie.cpp:
```diff
@@ -0,0 +1,470 @@
+    }
+
+    std::string revision;
+    CryptoPP::StringSource ss(header.revision.data(), sizeof(header.revision), true,
```
There's an ArrayToString function that doesn't depend on CryptoPP in Common.

Subv commented on src/core/movie.cpp:
```diff
@@ -0,0 +1,470 @@
+#pragma pack(pop)
+
+static PlayMode play_mode = PlayMode::None;
+static std::vector<u8> temp_input;
```
Surely you can come up with a better name for this variable

danzel replied:
But dolphin.... https://github.com/dolphin-emu/dolphin/blob/master/Source/Core/Core/Movie.cpp#L73
hahaha will rename :)

jroweboy commented on src/core/movie.cpp:
```diff
@@ -0,0 +1,470 @@
+constexpr std::array<u8, 4> header_magic_bytes{{'C', 'T', 'M', 0x1B}};
+
+#pragma pack(push, 1)
+struct CTMHeader {
```
Is there any advantage to having a fixed size header instead of a variable sized header with a fixed "start" location. So for example:

```cpp
std::array<u8, 20> revision;
u32 start_offset;
```

then we shouldn't have to worry as much about running out of header space since each movie file could potentially have a different start location.

(Note this isn't much of a review concern as it is a review comment)

danzel replied:
Nah, could change this. No big advantage either way IMHO.
Variable is a bit more complicated, fixed wastes a little space.

jroweboy commented on src/core/movie.cpp:
```diff
@@ -0,0 +1,470 @@
+
+#pragma pack(push, 1)
+struct CTMHeader {
+    std::array<u8, 4> filetype;  // Unique Identifier (always "CTM"0x1B)
```
triple slash to make it a doc comment (and maybe make them a little less silly :D)

danzel commented:
Implemented latest feedback and resquashed.
Individual latest commit is here https://github.com/danzel/citra/commit/06f98fe2c6b5244a25cc93d9878c633fa6597194

bunnei commented:
@danzel thanks for proposing this change. This PR is getting a bit stale - we need to decide if we'd like to get this merged, or if it's not something that is of value to us and should be abandoned. Personally, I'm not really familiar with Dolphin's (or other emulator's) ability to record game inputs, and what it might be useful for. Can other folks on the team comment on this, and/or review this PR if this feature is something we should adopt?

danzel commented:
It's has been suggested I copy over some propoganda from discord, so...

![image](https://user-images.githubusercontent.com/393086/34031271-7234e712-e1d5-11e7-9a99-6a1f1e0d99dc.png)

![image](https://user-images.githubusercontent.com/393086/34031275-75c2aeaa-e1d5-11e7-8fb8-d340bb071953.png)
The implementation is as close to dolphin as I could get it

![image](https://user-images.githubusercontent.com/393086/34031292-82713144-e1d5-11e7-86ac-db1c9579aec5.png)
Absolutely something we will need to think about once we get to that part, a bit outside of the scope of this PR however.

![image](https://user-images.githubusercontent.com/393086/34031298-8650e4ee-e1d5-11e7-882c-e4e7799d0e7f.png)

![image](https://user-images.githubusercontent.com/393086/34031241-49a52b22-e1d5-11e7-9c24-70578109c201.png)
Currently we record based on how often hardware is polled.
We can change this in the future if it turns out to not be a good way to do it (We can't do it frame-based as we can update inputs in the middle of a frame, but we could record timestamps instead).

![image](https://user-images.githubusercontent.com/393086/34031027-6563ef16-e1d4-11e7-987d-f0e20987f273.png)

I'll look at rebasing this again-again this weekend.

danzel commented:
Some notes for comparison with dolphin:

[Movie.cpp](https://github.com/dolphin-emu/dolphin/blob/master/Source/Core/Core/Movie.cpp)
Our Movie.cpp is based on the same idea of dolphins. We have a big buffer we write to (recording) or read from (playback).
Dolphins file is quite a bit bigger because it has methods in their for showing the current inputs and a bunch of stuff we don't need yet (they probably shouldn't be in that file imho)

Every time the emulator updates the state of the controller inputs (GC / Wiimote) it does the normal update (ask SDL for the buttons held etc), then calls in to the movie system so it can (when enabled) replace the state with the recorded one, or copy it in to the recording.

These are hooked in each place an input is read:
https://github.com/dolphin-emu/dolphin/blob/d89698898abeda018dc2b8896f00bc46f036f36b/Source/Core/Core/HW/WiimoteEmu/WiimoteEmu.cpp#L810
https://github.com/dolphin-emu/dolphin/blob/38f44b2839d1cb3792f1bf517efe4a092933c43e/Source/Core/Core/HW/SI/SI_DeviceGCController.cpp#L163

They support network inputs and some other thigs which we don't, so again theirs is a bit bigger.
However we have multiple input types (Pad, Touch Screen, C-Pad, Circle Pad, Accel/Gyro iirc), so we need to handle recording all of these different types of inputs.

B3n30 commented on src/core/movie.cpp:
```diff
@@ -0,0 +1,469 @@
+
+static void Play(Service::HID::PadState& pad_state, s16& circle_pad_x, s16& circle_pad_y) {
+    ControllerState s;
+    std::memcpy(&s, &recorded_input[current_byte], sizeof(ControllerState));
```
Assert or check for CheckInputEnd before copy?

danzel replied:
We check and update after every Play call (see Handle template), but adding an assert before calling Play too. (Could happen if given an invalid file to playback)

B3n30 commented on src/core/movie.cpp:
```diff
@@ -0,0 +1,469 @@
+
+    SaveMovie();
+
+    play_mode = PlayMode::None;
```
Those values should probably get reseted independently of IsRecordingInput(). So move that in front of the check

danzel replied:
:+1:

danzel commented on src/citra/citra.cpp:
```diff
@@ -121,11 +131,17 @@ int main(int argc, char** argv) {
     }
 
+    if (!movie_record.empty() && !movie_play.empty()) {
+        LOG_CRITICAL(Frontend, "Cannot both play and record a movie");
```
Yep!

danzel commented:
Rebased and fixed things based on B3n30s feedback above.

jroweboy commented:
After several months, more than enough reviews, no strong opposition, many changes made in the name of bikeshedding, lets just click the merge button already!
